compiler/src/iree/compiler/Dialect/HAL/Target/VulkanSPIRV/VulkanSPIRVTarget.cpp

```diff
@@ -46,35 +46,66 @@ VulkanSPIRVTargetOptions getVulkanSPIRVTargetOptionsFromFlags() {
   // llvm::cl::OptionCategory halVulkanSPIRVOptionsCategory(
   //     "IREE Vulkan/SPIR-V backend options");
 
-  static llvm::cl::opt<std::string> clVulkanTargetTriple(
+  static llvm::cl::list<std::string> clVulkanTargetTriples{
       "iree-vulkan-target-triple",
       llvm::cl::desc(
           "Vulkan target triple controlling the SPIR-V environment."),
-      llvm::cl::init("unknown-unknown-unknown"));
+  };
 
-  static llvm::cl::opt<std::string> clVulkanTargetEnv(
+  static llvm::cl::list<std::string> clVulkanTargetEnvs{
       "iree-vulkan-target-env",
       llvm::cl::desc(
           "Vulkan target environment as #vk.target_env attribute assembly."),
-      llvm::cl::init(""));
+  };
 
   static llvm::cl::opt<bool> clVulkanIndirectBindings(
       "iree-vulkan-experimental-indirect-bindings",
       llvm::cl::desc("Force indirect bindings for all generated dispatches."),
       llvm::cl::init(false));
 
   VulkanSPIRVTargetOptions targetOptions;
-  targetOptions.targetEnv = clVulkanTargetEnv;
-  targetOptions.targetTriple = clVulkanTargetTriple;
+  targetOptions.targetEnvs = std::vector<std::string>(
+      clVulkanTargetEnvs.begin(), clVulkanTargetEnvs.end());
+  targetOptions.targetTriples = std::vector<std::string>(
+      clVulkanTargetTriples.begin(), clVulkanTargetTriples.end());
+
+  std::vector<bool> isEnv;
```

Comment:
Do we need such complicated logic to handle the mix of target triples and target environments? Do you anticipate a need? IMO we should just allow either as everything target triples, or everything as target environments. We can always scale to support the mix when the needs really come (I don't see it right now); so in another way, avoid introducing complicated mechanisms until it's really needed.

Reply:
My thinking was that it would be common to want to specify some minimal target (not based on any specific architecture) as a `vk.target_env`, and then specify the maximum target with a `target-triple`.

Reply:
If you insist on making it all or nothing for now, that's fine, but the way I'm envisioning early versions of this, having the ability to do both sounds very practical to me (e.g. specify some minimal environment that covers most recent desktop GPUs between NVidia and AMD) + a few of the target triples for the cards you really care about (that maybe have more advanced extensions like cooperative matrix). One other way we could do it for now is to always prioritize the `target-triple` over the `target_env` because target triples are almost always going to be as specific as possible (and thus should be prioritized).

Reply:
(the counter to `target-triple` always > `target_env` would be when compiling for both mobile and desktop simultaneously, but we don't have to optimize for that right out of the gate).

Reply:
Yeah, mostly about the implementation complexity and mental overhead. As we discussed, in the long run we want to have the minimal target capabilities deduction anchored on the particular root op when fanning out variant ops--at that time we know concretely and precisely the minimal target capabilities needed; and each different root op would need different sets there. Specifying the minimal target capabilities via the command line is likely intermediate, if ever needed at a. Don't need to spend too much efforts over polishing it; just get the straightforward solution to carry us over the current stage. :)

Thinking a bit more, I'm fine of mixing the target triple and environment. They are essentially the same thing, with the target triple being a shorthand. Though I think we don't need to have priorities among them depending on relative positions or such; just treat everything as equal and a flat list. (Target triples are translated into target env strings under the hood anyway.) CodeGen towards all the environments equally would be fine.

Reply:
Makes sense. The way this PR works is it makes the target order equivalent to the order of arguments that the user passes in (e.g. if the user passes `target_env`, `target-triple`, `target_env`, the resulting arrayattr has the same order). So in a sense it is trying to flatten it based on the order the user specified (basically because the way that it works right now, target environments specified earlier on will be picked before later ones). So from the user's perspective, it just uses whatever order they specified through flags.

Reply:
Just to clarify, the priority has little to do with Codegen. Codegen will just see the list of targets and generate code for each one. The priority is for the host side, where if multiple environments match the current one, then it knows which one to use, and the way that manifests is just as an ordering in the resulting switch statement. So basically what I wanted here was to make the order the user specified in the flags match with the order in that switch statement, but also while being able to mix target triples with target envs. Just a convenience thing.

Reply:
I made it so that they both use the same list of strings, but kept the logic for reverse-engineering the flag order from the user. Now the helper for constructing the target attribute looks into the string to try and guess whether it is a target triple or a target env attribute. Let me know if this works for you.

Reply:
If this is still more complexity than needed, I can either switch to having one shared flag for target triples and target envs, or I can just do `executable_targets = [<all target triples>, <all target envs>]`. I didn't want to change the flag names because that will mess up a lot of off-hand scripts and potential downstream projects using them.

Reply:
Great! Yeah unifying `--iree-vulkan-target-triple` and `--iree-vulkan-target-env` into one option could be a way but don't want to break the whole world for now.